src/test/ledger/Invariants_test.cpp

```diff
@@ -798,6 +798,110 @@ class Invariants_test : public beast::unit_test::suite
+            {{"permissioned domain with no rules."}},
+            [](Account const& A1, Account const&, ApplyContext& ac) {
+                Keylet const pdKeylet = keylet::permissionedDomain(A1.id(), 10);
+                auto slePd = std::make_shared<SLE>(pdKeylet);
+                slePd->setAccountID(sfOwner, A1);
+                slePd->setFieldU32(sfSequence, 10);
+
+                ac.view().insert(slePd);
+                return true;
+            },
+            XRPAmount{},
+            STTx{ttPERMISSIONED_DOMAIN_SET, [](STObject& tx) {}},
+            {tecINVARIANT_FAILED, tecINVARIANT_FAILED});
+
+        testcase << "PermissionedDomain 2";
+        doInvariantCheck(
+            {{"permissioned domain bad credentials size 11"}},
+            [](Account const& A1, Account const& A2, ApplyContext& ac) {
+                Keylet const pdKeylet = keylet::permissionedDomain(A1.id(), 10);
+                auto slePd = std::make_shared<SLE>(pdKeylet);
+                slePd->setAccountID(sfOwner, A1);
+                slePd->setFieldU32(sfSequence, 10);
+
+                STArray credentials(sfAcceptedCredentials, 11);
+                for (std::size_t n = 0; n < 11; ++n)
+                {
+                    auto cred = STObject::makeInnerObject(sfCredential);
+                    cred.setAccountID(sfIssuer, A2);
+                    auto credType =
+                        std::string("cred_type") + std::to_string(n);
+                    cred.setFieldVL(
+                        sfCredentialType,
+                        Slice(credType.c_str(), credType.size()));
+                    credentials.push_back(std::move(cred));
+                }
+                slePd->setFieldArray(sfAcceptedCredentials, credentials);
+                ac.view().insert(slePd);
+                return true;
+            },
+            XRPAmount{},
+            STTx{ttPERMISSIONED_DOMAIN_SET, [](STObject& tx) {}},
+            {tecINVARIANT_FAILED, tecINVARIANT_FAILED});
+
+        testcase << "PermissionedDomain 3";
+        doInvariantCheck(
+            {{"permissioned domain credentials aren't sorted"}},
+            [](Account const& A1, Account const& A2, ApplyContext& ac) {
+                Keylet const pdKeylet = keylet::permissionedDomain(A1.id(), 10);
+                auto slePd = std::make_shared<SLE>(pdKeylet);
+                slePd->setAccountID(sfOwner, A1);
+                slePd->setFieldU32(sfSequence, 10);
+
+                STArray credentials(sfAcceptedCredentials, 11);
+                for (std::size_t n = 0; n < 2; ++n)
+                {
+                    auto cred = STObject::makeInnerObject(sfCredential);
+                    cred.setAccountID(sfIssuer, A2);
+                    auto credType =
+                        std::string("cred_type") + std::to_string(9 - n);
+                    cred.setFieldVL(
+                        sfCredentialType,
+                        Slice(credType.c_str(), credType.size()));
```

Comment:
I didn't fully understand the `reinterpret_cast` operation in the `Slice` constructor. Does it preserve the lexicographic order of the contents?

How can we prove that the order of credentials is not sorted after this operation? An explicit typecast into `std::uint8_t const*` might change the order isn't it ?

Reply:
No, just convert void to uint pointer, nothing will be sorted silently (there is no such ability in the c++ itself)

Reply:
I'm referring to the `makeSorted` function in `CredentialHelpers.cpp` file. It sorts the values based on `<AccountID, Slice>` pairs. 

After the below line:
`cred.setFieldVL(sfCredentialType, Slice(credType.c_str(), credType.size()));`

is the non-ordered nature of the input preserved?

I'm not able to visualize the output of the `Slice()` operation because its `std::uint_8t const*` pointers.

Reply:
It is  desc ordered as you can see from` std::string("cred_type") + std::to_string(9 - n);`
It will be 
```
cred_type9
cred_type8
```
Which is unordered for accepted credentials field.